File: LeetCodeProblemsGo/LeetCode2780MinimumIndexOfAValidSplit/LeetCode2780MinimumIndexOfAValidSplit.go
package leetcode2780minimumindexofavalidsplit

func minimumIndex(nums []int) int {
	dominantElement := nums[0]
	frequency := 1
	length := len(nums)
	if length == 1 {
		return -1
	}
	for i := 1; i < length; i++ {
		if nums[i] == dominantElement {
			frequency++
		} else {
			frequency--
		}
		if frequency == 0 {
			dominantElement = nums[i]
			frequency = 1
		}
	}
	dominantCount := 0
	for i := 0; i < length; i++ {
		if nums[i] == dominantElement {
			dominantCount++
		}
		nums[i] = dominantCount
	}
	for i := 0; i < length; i++ {
		if nums[i] <= (i+1)/2 {
			continue
		}
		if dominantCount-nums[i] <= (length-i-1)/2 {
			continue
		}
		return i
	}
	return -1
}

File: LeetCodeProblemsCMake/Problems/LeetCode2780MinimumIndexOfAValidSplit.cpp
//
// Created by saeed on 3/27/25.
//

#include "LeetCode2780MinimumIndexOfAValidSplit.h"

#include <unordered_map>
using namespace std;

int LeetCode2780MinimumIndexOfAValidSplit::minimumIndex(vector<int> &nums) {
    unordered_map<int, vector<int> > occurrences{};
    int length = static_cast<int>(nums.size());
    int dominantElement{};
    for (int i = 0; i < length; i++) {
        if (!occurrences.contains(nums[i])) {
            occurrences[nums[i]] = vector<int>{};
        }
        occurrences[nums[i]].push_back(i);
        if (occurrences[nums[i]].size() >= length / 2) {
            dominantElement = nums[i];
        }
    }
    int dominantCount = static_cast<int>(occurrences[dominantElement].size());
    int domCountBefore{};
    int elementsBefore{};
    int domCountAfter{};
    int elementsAfter{};

    for (int i = 0; i < dominantCount; i++) {
        domCountBefore = i + 1;
        elementsBefore = occurrences[dominantElement][i] + 1;
        domCountAfter = dominantCount - domCountBefore;
        elementsAfter = length - elementsBefore;
        if (domCountBefore <= elementsBefore / 2)
            continue;
        if (domCountAfter <= elementsAfter / 2)
            continue;
        return occurrences[dominantElement][i];
    }
    return -1;
}
